ttnn/cpp/ttnn/tensor/tensor_layout.cpp

```diff
@@ -0,0 +1,358 @@
+        values[rank - 1] = legacyPaddedShape[rank - 1];
+    }
+    if(rank > 2) {
+        values[rank - 2] = legacyPaddedShape[rank - 2];
+    }
+    for (size_t i = rank - 3; i >= 0; i--) {
+        values[i] = legacyPaddedShape[i] * values[i + 1];
+    }
+
+    Alignment result(values);
+    return result;
+}
+}
+
+// Private constructor to create TensorLayout from LegacyPaddedShape
+TensorLayout::TensorLayout(DataType dataType, Layout layout, const MemoryConfig& memoryConfig, const ttnn::SimpleShape& legacyPaddedShape)
+    : TensorLayout(dataType, layout, memoryConfig, legacyPaddedShapeToAlignment(legacyPaddedShape)) {
+    mLegacyPaddedShape = legacyPaddedShape;
+}
+
+TensorLayout TensorLayout::fromLegacyPaddeShape(DataType dataType, Layout layout, const MemoryConfig& memoryConfig, const ttnn::SimpleShape& legacyPaddedShape) {
+    return TensorLayout(dataType, layout, memoryConfig, legacyPaddedShape);
+}
+
+void TensorLayout::initializeAlignment() {
+    if(mAlignment.size() != 0)
+        return;
+
+    switch(mLayout) {
+        case Layout::ROW_MAJOR: {
+            mAlignment = Alignment({sizeof(uint32_t) % element_size_bytes()});
+            break;
+        }
+        case Layout::TILE: {
+            mAlignment = Alignment({mTileSize.height(), mTileSize.width()});
+            break;
+        }
+        case Layout::INVALID:
+            TT_THROW("Invalid Layout");
+    }
+}
+
+void TensorLayout::validateCustomAlignment() const
+{
+    if(mAlignment.size() == 0)
+        return;
+
+    switch(mLayout) {
+        case Layout::ROW_MAJOR: {
+            if(mAlignment.size() > 0) {
+                auto widthAlignment = mAlignment[-1];
+                auto element_size = element_size_bytes();
+                auto page_alignment = sizeof(uint32_t) % element_size;
+                TT_FATAL(widthAlignment % page_alignment == 0,
+                "Wrong custom Tensor Layout alignment {}. For Row Major layout with element size {}bytes the innermost dimension must align to {}. This is because Buffer data is packes as uint32_t (4 bytes).",
+                    mAlignment,
+                    element_size,
+                    page_alignment);
+            }
+            break;
+        }
+        case Layout::TILE: {
+            if(mAlignment.size() > 1) {
+                auto widthAlignment = mAlignment[-1];
+                TT_FATAL(widthAlignment % mTileSize.width() == 0,
+                "Wrong custom Tensor Layout alignment {}. For Tile layout innermost dimension should be multiple of tile width {}.", mAlignment, mTileSize.width());
+            }
+            if(mAlignment.size() > 2) {
+                auto heightAlignment = mAlignment[-2];
+                TT_FATAL(heightAlignment % mTileSize.height() == 0,
+                "Wrong custom Tensor Layout alignment {}. For Tile layout second innermost dimension should be multiple of tile height {}.", mAlignment, mTileSize.height());
+            }
+            break;
```

Comment:
Shouldn't we require that `mAlignment.size()` must be `>= 2` for tile layout?

Reply:
fair. started this when I assumed that user might provide an empty one, but its now initialized so its fair